change combine_grid API

hrnr opened the pull request:
Hi again,

I have another changelist for parts I use.

The current API for combine grids takes shared_ptr, which is not only theoreticaly and semanticaly wrong as not ownership is passed, but also prevents you completely to safely manage your occupancy grids without shared_ptrs.

I added 2 backward compatible functions using shared pointers.

Cheers,
Jiri

hrnr committed:
change combine_grid API

do not take share_ptr, take normal pointers only. No ownership is tranfered so using shared_ptr here is WRONG! It does not allow caller to manage occupancy grids in other ways than shared_ptr (ie. you can't use occupancy grids allocated on stack at all).

return merged grid via reference. This prevents copying in <C++11 without move semantics.

hrnr committed:
combine_grids: backward compatible functions

add compatibility functions to support old shared_ptr api

efernandez commented:
@hrnr Could you provide a use case (example code) using the new methods?

What if someone wants to have the OGMs on a vector owning them, so they're destroyed automatically?

hrnr commented:
I see your point. I have choosen this aproach so it would be possible to implement backward compatible functions. But you're right it won't work in C++ this way.

The obvious proper way is to take (begin, end) iterators, but then it might be tricky to implement backward compatible functions. I'll give it a try. With std::reference_wrapper it should be possible.

btw Do you even care for backward compatibility here?

hrnr committed:
change occupancy grid API

accept iterator pair instead a vector - as do all algorithms operating on containers. This seems to me as most universal aproach.

Accepting iterators requires templates. Required code moved to header.

Compatibility functions reimplemented using boost::reference_wrapper. This trick is also useful if user wants to provide `view` as it was possible with pointer semantics. So supporting reference_wrapper (few additional lines thats trigger conversions) is useful even after compat functions will be removed.

hrnr committed:
remove ancient gcc hacks

hrnr commented:
Ok. I have changed that to accept iterators and used some trickery for compat functions. This seems to me as the most generic concept, same as <algorithm> functions.

Some code needed to go to header so it's now much biger, but this is C++, right?

hrnr commented:
Hi,

what do you think about latest changes?

The new API follows standad C++ semantics and as a bonus I was able to implement compat functions. This API is much more flexible for my needs.

efernandez commented:
FYI @afakihcpr

hrnr commented:
I'm no longer interested in this. I have realized I can do the same job with ~100 SLOC. But thanks for maintaining this legacy code.
